packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -62,11 +46,192 @@ class PlaceholderSpanIndexSemanticsTag extends SemanticsTag {
   int get hashCode => Object.hash(PlaceholderSpanIndexSemanticsTag, index);
 }
 
+/// Parent data used by [RenderParagraph] and [RenderEditable] to annotate
+/// inline widgets with.
+class TextParentData extends ParentData with ContainerParentDataMixin<RenderBox> {
+  /// The offset at which to paint the child in the parent's coordinate system.
+  ///
+  /// A `null` value indicates this inline widget is not laid out. For instance,
+  /// when the inline widget has never been laid out, or the inline widget is
+  /// ellipsized away.
+  Offset? get offset => _offset;
+  Offset? _offset;
+
+  /// The [WidgetSpan] associated with this inline widget.
+  ///
+  /// This field is usually set by a [ParentDataWidget], and is typically not
+  /// null when `performLayout` is called.
+  PlaceholderSpan? span;
+
+  @override
+  void detach() {
+    span = null;
+    _offset = null;
+    super.detach();
+  }
+
+  @override
+  String toString() =>'widget: $span, ${offset == null ? "not laid out" : "offset: $offset"}';
+}
+
+/// A mixin that provides useful default behaviors for text [RenderBox]es
```

Comment:
what are the defaults?

Reply:
The default layout/paint/hitTest implementations I assume. I followed this class: https://master-api.flutter.dev/flutter/rendering/RenderBoxContainerDefaultsMixin-mixin.html

Reply:
Added documentation for when to call each default method.